src/nnet3/nnet-test-utils.cc

```diff
@@ -1026,12 +1026,21 @@ static void GenerateRandomComponentConfig(std::string *component_type,
     case 24: {
       *component_type = "RepeatedAffineComponent";
       int32 num_repeats = RandInt(1, 50),
-          input_dim = num_repeats * RandInt(1, 15),
-          output_dim = num_repeats * RandInt(1, 15);
```

Comment:
I think I see what might be the problem.  The following code:
      int32 size_as_vector =
          (linear_params_.NumRows() - 1) * linear_params_.Stride() +
          linear_params_.NumCols(),
          stride_as_matrix = linear_params_.NumRows() *
linear_params_.Stride();
where it says linear_params_.Stride(), needs to refer to
linear_params_deriv_repeated.Stride().
They are usually but not always the same.
Dan




On Thu, Dec 31, 2015 at 11:31 PM, Daniel Galvez <notifications@github.com>
wrote:

> In src/nnet3/nnet-test-utils.cc
> <https://github.com/kaldi-asr/kaldi/pull/427#discussion_r48676612>:
>
> > @@ -1026,12 +1026,21 @@ static void GenerateRandomComponentConfig(std::string *component_type,
> >      case 24: {
> >        *component_type = "RepeatedAffineComponent";
> >        int32 num_repeats = RandInt(1, 50),
> > -          input_dim = num_repeats * RandInt(1, 15),
> > -          output_dim = num_repeats * RandInt(1, 15);
>
> Okay, I have confirmed that the tests fail only RepeatedAffineComponent
> (NOT BlockAffineComponent. That looks fine.), when input_dim / num_repeats
> and output_dim / num_repeats are equal to 1. Even then, it is not
> guaranteed to fail, however. It can also fail for other small values.
>
> I suspect the reason has to do with striding. I'm worried your hack to
> view the vector as a matrix assumes that does not take into account the
> effect of extra padding used between rows for the sake of alignment. A
> padding issue exactly like the kind of thing that would come up only with
> very small dimensions. I don't know CUDA, but the documentation
> <http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g32bd7a39135594788a542ae72217775c>
> mentions that padding is used in allocations.
>
> I'll test this hypothesis soon, hopefully tomorrow.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/427/files#r48676612>.
>